src/garage/envs/environment.py

```diff
@@ -0,0 +1,188 @@
+
+        Returns:
+            akro.Space: The envionrment specification.
+        """
+
+    @property
+    @abc.abstractmethod
+    def spec(self):
+        """Return the environment specification.
+
+        Returns:
+            garage.envs.env_spec.EnvSpec: The envionrment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def render_modes(self):
+        """Return the list of supported render modes.
+
+        Returns:
+            list: A list of string representing the supported render modes.
+                  See render() for a list of modes.
+
+        """
+
+    @abc.abstractmethod
+    def reset(self):
+        """Resets the environment and returns the first `TimeStep` of sequence.
+
+        Returns:
+            garage.TimeStep: The first TimeStep.
+                Note that this `TimeStep` will have attributes `observation`,
+                `action`, `reward` as None as they are undefined.
+
+        """
+
+    @abc.abstractmethod
+    def step(self, action):
+        """Steps the environment using to the action and returns a `TimeStep`.
+
+        If the environment returned the last time step of a sequence (either
+        because of terminal or time out) at the previous step, this call to
+        `step()` will start a new sequence and `action` will be ignored.
+
+        This method will also start a new sequence if called after the
+        environment has been constructed and `reset()` has not been called.
+        Again, in this case `action` will be ignored.
+
+        If possible, update the visualization display as well.
+
+        Args:
+            action (np.ndarray): A NumPy array, or a nested dict, list or tuple
+                of arrays corresponding to `action_spec()`.
+
+        Returns:
+            garage.TimeStep: The TimeStep after the action.
+
+        """
+
+    @abc.abstractmethod
+    def render(self, mode):
+        """Renders the environment.
+
+        The set of supported modes varies per environment. By convention,
+        if mode is:
+
+        * rgb_array: Return an numpy.ndarray with shape (x, y, 3),
+          representing RGB values for an x-by-y pixel image, suitable
+          for turning into a video.
+        * ansi: Return a string (str) or StringIO.StringIO containing a
+          terminal-style text representation. The text can include newlines
+          and ANSI escape sequences (e.g. for colors).
+
+        Make sure that your class's `render_modes` includes the list of
+        supported modes.
+
+        For example:
+
+            class MyEnv(Env):
+                def render_modes(self):
+                    return ['rgb_array', 'ansi']
+
+                def render(self, mode):
+                    if mode == 'rgb_array':
+                        return np.array(...) # return RGB frame for video
+                    elif mode == 'ansi':
+                        ... # return text output
+                    else:
```

Comment:
i'm not sure what the else case here makes clear your intention?

I think the idiomatic thing to do here (and what we should encourage implementers to do) is explicitly raise a `ValueError` which says (1) which mode you got and (2) which modes you support.